### DISCUSSION TITLE
Deprecated status for target languages

### DISCUSSION THREAD

**Discussion by wsfulton** [swig/swig#3038] [General]
I've added a new 'Deprecated' status into the docs in 447cd4dd14fc11664c6b2ecbade91cabe5f3dffd for target languages. The idea is that this formalises what we are doing for deprecating MzScheme in #2830 and a refinement of #1447 for the removal of those target languages.

Please suggest refinements, improvements or objections to this new process of removal of target languages here.

**Comment by erezgeva** [comment]
Post a call for help in the project maintaining the target language.
Developers that are involved in the projects are good candidate for help.

**Reply by ojwb** [reply]
Some thoughts:

> if recent versions cannot successfully run the test-suite and examples on the Github Actions Continuous Integration platform

This seems overly specific - the key thing is we have CI rather than where it's hosted.  There are potentially practical issues with having our CI split across systems (especially that they seem to have standardised on YAML as the file format, but not on a compatible format within the YAML), but really the biggest problem with appveyor is that it seems to run one job at a time rather than that it's not part of github actions.

> for removal from SWIG in a subsequent release (usually one SWIG release).

I think that should probably be "usually the next x.y.0 SWIG release" or something to similar effect - if we need a 4.3.1 point release to fix a regression in 4.3.0 it would seem surprising to remove mzscheme at that point.  Perhaps we should also specify that the deprecation ought to happen in a .0 release too (so we shouldn't deprecate in 4.3.1 and remove in 4.4.0, because then upgrading from 4.3.0 to 4.3.1 would change the status of a target language which again seems surprising).

We don't actually seem to formally define what the three version components mean that I can see, beyond this in the definition of `Supported` status:

> Each point release must aim to be fully backwards compatible. A point release version is the 3rd version digit, so each of the x.y.* versions should be backwards compatible. Backwards compatibility breakages can occur in a new major or minor version if absolutely necessary and if documented. A major or minor version is the first or second digit in the three digit version.

> Post a call for help in the project maintaining the target language.
> Developers that are involved in the projects are good candidate for help.

We really want someone to step up to act as maintainer going forwards, but those developers are likely already busy and not directly motivated to maintain SWIG's support.  Someone stepping in to add support for racket 8.0 and then disappearing right away means the burden of maintenance falls on core SWIG developers who are already overstretched.

A better option is really someone actively using SWIG to generate wrappers in the target language as they're directly motivated to keep support working and to improve it.

**Reply by erezgeva** [reply]
> but really the biggest problem with appveyor is that it seems to run one job at a time rather than that it's not part of github actions.

We agree.
We already reduce the tests on AppVeyor to 3.
And we plan to add Python with MingW on GitHub action.

> A better option is really someone actively using SWIG to generate wrappers in the target language as they're directly motivated to keep support working and to improve it.

This sound as a better option.
But as it may be difficult to find a developer that actually understand the C API.
Perhaps we can split. One developer that perform the migration and another that maintain it after.
One with a good knowledge on the target language C API.
And another with the will to use SWIG.

**Reply by ojwb** [reply]
> We agree.
> We already reduce the tests on AppVeyor to 3.
> And we plan to add Python with MingW on GitHub action.

Right, but my point here is really that if a target language had CI but only on appveyor (or only on some other non-GHA CI system) that doesn't in itself seem like enough of a problem to justify deprecating and removing it.  It's a hypothetical concern at the moment, but it's best if the policy as written reflects closely what we would actually do.

**Comment by erezgeva** [comment]
Looking here: https://www.cs.umb.edu/~ryanc/tmp/swig/Racket.html
Gave me another idea.
This web is under 'Ryan Culpepper' home, when he was in 
[University of Massachusetts Boston, Department of COMPUTER SCIENCE](https://www.cs.umb.edu/).

I know there were some computer science professors involved in the project in the past.
University students sound a good pool of potential develops.
